src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -437,33 +454,131 @@ public static unsafe void AddScalarU(float scalar, Span<float> dst)
+
+                        remainder = length % 4;
+
+                        for (float* pEnd = pDstCurrent + (length - remainder); pDstCurrent < pEnd; pDstCurrent += 4)
+                        {
+                            if (Avx.IsSupported)
+                            {
+                                // The JIT will only fold away unaligned loads due to the semantics behind
+                                // the VEX-encoding of the memory operand for `ins xmm, xmm, [mem]`. Since
+                                // modern hardware has unaligned loads that are as fast as aligned loads,
+                                // when it doesn't cross a cache-line/page boundary, we will just assert
+                                // that the alignment is correct and allow for the more-efficient codegen.
+
+                                Contracts.Assert(((nuint)(pDstCurrent) % 16) == 0);
+                                Vector128<float> temp = Sse.LoadVector128(pDstCurrent);
+                                temp = Sse.Multiply(scaleVector128, temp);
+                                Sse.Store(pDstCurrent, temp);
+                            }
+                            else
+                            {
+                                // If we aren't using the VEX-encoding, then the reverse is true and the JIT
+                                // will only fold away aligned loads (due to semantics of the legacy encoding).
+                                // We don't need an assert, since the instruction will throw for unaligned inputs.
+                                Vector128<float> temp = Sse.LoadAlignedVector128(pDstCurrent);
+                                temp = Sse.Multiply(scaleVector128, temp);
+                                Sse.Store(pDstCurrent, temp);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // Handle the "worst-case" scenario, which is when we have 4-8 elements and the input is not
+                        // 128-bit aligned. This means we can't do any aligned loads and will just end up doing two
+                        // unaligned loads where we mask the input each time.
+                        remainder = length;
+                    }
+
+                    if (remainder != 0)
+                    {
+                        // Handle any trailing elements that don't fit into a 128-bit block by moving back so that the next
+                        // unaligned load will read to the end of the array and then mask out any elements already processed
+
+                        pDstCurrent -= (4 - remainder);
+
+                        Vector128<float> result = Sse.LoadVector128(pDstCurrent);
+
+                        Vector128<float> trailingMask = Sse.LoadVector128(((float*)(pTrailingAlignmentMask)) + (remainder * 4));
```

Comment:
maybe swap this and the next line so you're consistently doing leading before trailing

Reply:
I think the current way makes more sense.

The first case had leading as primary and trailing as secondary.
While this case has trailing as primary and leading as secondary.